### ISSUE TITLE
check for $BUILD_PREFIX 

### ISSUE THREAD

**Issue by minrk** [conda/conda-build#2995]
A bit related to: #2677

### Actual Behavior

References to $BUILD_PREFIX are preserved, unmodified, in built packages

### Expected Behavior

Warnings and/or replacements of $BUILD_PREFIX somehow, because references to $BUILD_PREFIX could result in non-working packages if the references are used in any way.

It would be useful to check for $BUILD_PREFIX in the results and at least warn about these references, because the recipe may want or need to change something. A step further would be a replacement along the lines of what's done with $PREFIX (either the same as $PREFIX, or rewrite to an opaque "this_was_the_build_env" path that clearly shouldn't work).

Perhaps the simplest solution is to warn about $BUILD_PREFIX and point recipe-authors to solution:

- maybe there's a dependency that should be in host instead of build?
- `build.merge_build_and_host` as a possible solution

A strict flag or `conda check` to error if references are found could be useful as well.

### Steps to Reproduce

build a recipe that records a reference to $BUILD_PREFIX, e.g. by recording the path to the compiler, build prefix include dirs, etc.. petsc is one such package that over-records absolute paths of things used during build.

An example build.sh that would trigger such behavior:

```bash
# build.sh
echo "$BUILD_PREFIX" > $PREFIX/share/build-prefix
```

##### `Output of conda info`

```

     active environment : None
            shell level : 0
       user config file : /Users/benjaminrk/.condarc
 populated config files : /Users/benjaminrk/conda/.condarc
                          /Users/benjaminrk/.condarc
          conda version : 4.5.4
    conda-build version : 3.10.9+16.g15607ea0.dirty
         python version : 3.6.5.final.0
       base environment : /Users/benjaminrk/conda  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/osx-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/osx-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/osx-64
                          https://repo.anaconda.com/pkgs/pro/noarch
          package cache : /Users/benjaminrk/conda/pkgs
                          /Users/benjaminrk/.conda/pkgs
       envs directories : /Users/benjaminrk/conda/envs
                          /Users/benjaminrk/.conda/envs
               platform : osx-64
             user-agent : conda/4.5.4 requests/2.19.1 CPython/3.6.5 Darwin/17.6.0 OSX/10.13.5
                UID:GID : 501:20
             netrc file : /Users/benjaminrk/.netrc
           offline mode : False
```

**Comment by mingwandroid** [comment]
As a workaround for this in a few packages I've done things such as:

`export CC="$(dirname ${CC})"`

.. near the top of build.sh. Not pretty!

**Comment by minrk** [comment]
Yeah, for the most common cases, I think that works. My latest example, petsc seems to *unconditionally* record the absolute path of the python interpreter (a build-only dependency), so I had to find those in the results and `sed` them to `python2`, otherwise downstream packages (slepc) would load petsc's "configure me like petsc" and try to execute `petsc_work/_build_env/bin/python` which doesn't exist.

I realize conda-build *fixing* all this in an automatic way is not really feasible. I'm mostly after warnings so that I can see the problem in the petsc build, rather than waiting to discover them in downstream packages.

It also means that I really want to be running `conda build <test-recipe>` as one of my tests, to verify that it works as intended. I'm not sure at what level a "test_build_recipes" belongs, but it would help me have confidence in my dylib recipes. In particular, I now need to test that I can build things that link my packages with a `build_env`, not just having compilers in the test env, which is no longer representative of a downstream conda build.

**Comment by mingwandroid** [comment]
Just adding a test that tries to use the software will exercise the path you need. conda-build moves build environments out of the way to make sure they don't get found during test.

We're also looking into adding downstream package test capability. Not sure where that is on the radar though.

**Comment by timsnyder** [comment]
@mingwandroid it also looks like `conda build` 3.18.9 is moving the 'host' environment out of the way but leaving the 'build' environment in place causing my test to pass when it shouldn't.  I'll file a separate issue for that (looking for one first).

**Comment by jaimergp** [comment]
It's still an issue.
